File: Algorithms/Dynamic Programing/fibonacci.py
calculations = 0


def fibonacci_recursive(n):
    if n < 2:
        return n

    global calculations
    calculations += 1
    return fibonacci_recursive(n-1) + fibonacci_recursive(n-2)


def memoized_fibonacci():
    cache = {}

    def fib(n):
        if cache.get(n):
            return cache[n]
        else:
            if n < 2:
                return n
            else:
                global calculations
                calculations += 1
                result = fib(n-1) + fib(n-2)
                cache[n] = result
                return result

    return fib


faster_fibonacci = memoized_fibonacci()
fib_large_num = faster_fibonacci(700)

print(fib_large_num)
print(f"It took {calculations} calculations")

File: Algorithms/Dynamic Programing/fibonacci.js
let calculations = 0;

function fibonacciRecursive(n) {
  if (n < 2) {
    return n;
  }

  return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

function memoizedFibonacci() {
  const cache = {};

  return function fib(n) {
    if (n in cache) {
      return cache[n];
    } else {
      if (n < 2) {
        return n;
      } else {
        calculations++;
        const result = fib(n - 1) + fib(n - 1);
        cache[n] = result;
      }
    }
  };
}

const fasterFibonacci = memoizedFibonacci();
const fibLargeNum = fasterFibonacci(9665);

console.log(fibLargeNum);
console.log(`It took ${calculations} calculations`);
